onnxruntime/test/util/include/inference_session_wrapper.h

```diff
@@ -20,6 +20,10 @@ class InferenceSessionWrapper : public InferenceSession {
     return model_->MainGraph();
   }
 
+  Graph& GetMutableGraph() const {
```

Comment:
I don't think you need to expose a mutable Graph. Resolve is called as part of model loading, so it shouldn't be necessary to call it again.

Reply:
The call to `resolve()` is to make sure that we recompute the shape inference data after the EP does the fusing. Unless I'm misunderstanding, in the model loading process all shape inference data will be computed _before_ the EP does the fusing (unless there are level 3 graph transforms which are run subsequently) – this was the whole motivating reason for this PR. So if we don't manually clear out the shape inference info, then we wouldn't be able to distinguish between shape inference performed by the fused node and shape inference info that was present prior to fusing.

Reply:
Graph::Resolve is called if the partitioning changes the Graph. If you need to do a manual call to Graph::Resolve in the test to make it work, that would imply it wouldn't work in normal usage.

Reply:
Under normal usage wouldn't the call to `graph::resolve` implicitly be done by any level-3 transformer that had been registered.

Reply:
Resolve could be called after a level 3 transformer changed the graph, or after there was a fusion during partitioning that changed the graph. Either way, if something changes the graph, a call to Graph::Resolve should happen prior to InferenceSession::Initialize completing. 

To validate that Resolve has been called, Graph::GraphResolveNeeded can be used (and a call to Graph::Resolve will do nothing unless GraphResolveNeeded returns true).

Will the test detect if the shape inferencing function is not called though? The shape info from prior to partitioning will still be present for the 'M' output won't it? If so, you could add a way for the FuseExecutionProvider to track that (no mutable Graph needed, assuming the Graph::Resolve is happening as expected), or you could manually remove the shape info for the 'M' output (would need to call NodeArg::ClearShape as well as Graph::SetResolveNeeded).

Reply:
Ah you're right, thanks for the info – I was under the impression that a call to `Resolve()` would _always_ recompute _all_ shape-info, but as you mentioned this is not the case.

> or you could manually remove the shape info
I've updated the test to do so